cli/manage.go

```diff
@@ -132,81 +140,71 @@ func getDiscoveryOpt(c *cli.Context) map[string]string {
 	return options
 }
 
-func getCandidateAndFollower(discovery discovery.Backend, addr string, leaderTTL time.Duration) (*leadership.Candidate, *leadership.Follower) {
-	kvDiscovery, ok := discovery.(*kvdiscovery.Discovery)
-	if !ok {
-		log.Fatal("Leader election is only supported with consul, etcd and zookeeper discovery.")
-	}
-	client := kvDiscovery.Store()
-	p := path.Join(kvDiscovery.Prefix(), leaderElectionPath)
-
-	candidate := leadership.NewCandidate(client, p, addr, leaderTTL)
-	follower := leadership.NewFollower(client, p)
-	return candidate, follower
-}
-
-func setupReplication(c *cli.Context, cluster cluster.Cluster, server *api.Server, candidate *leadership.Candidate, follower *leadership.Follower, addr string, tlsConfig *tls.Config) {
-	primary := api.NewPrimary(cluster, tlsConfig, &statusHandler{cluster, candidate, follower}, c.GlobalBool("debug"), c.Bool("cors"))
+func setupReplication(c *cli.Context, cluster cluster.Cluster, client store.Store, keyPath string, server *api.Server, candidate *leadership.Candidate, addr string, tlsConfig *tls.Config) {
+	handler := &statusHandler{cluster, candidate, nil, addr}
+	primary := api.NewPrimary(cluster, tlsConfig, handler, c.GlobalBool("debug"), c.Bool("cors"))
 	replica := api.NewReplica(primary, tlsConfig)
 
 	go func() {
 		for {
-			run(cluster, candidate, server, primary, replica)
-			time.Sleep(defaultRecoverTime)
-		}
-	}()
-
-	go func() {
-		for {
-			follow(follower, replica, addr)
-			time.Sleep(defaultRecoverTime)
+			err := run(cluster, client, keyPath, handler, server, primary, replica, addr)
+			if err != nil {
+				time.Sleep(defaultRecoverTime)
+			}
 		}
 	}()
 
 	server.SetHandler(primary)
 }
 
-func run(cl cluster.Cluster, candidate *leadership.Candidate, server *api.Server, primary *mux.Router, replica *api.Replica) {
-	electedCh, errCh := candidate.RunForElection()
+func run(cl cluster.Cluster, client store.Store, keyPath string, handler *statusHandler, server *api.Server, primary *mux.Router, replica *api.Replica, addr string) error {
+	candidate := handler.candidate
+	electedCh, candidateErrCh := candidate.RunForElection()
+
+	follower := leadership.NewFollower(client, keyPath)
+	defer follower.Stop()
+	handler.follower = follower
+	leaderCh, followerErrCh := follower.FollowElection()
+
 	var watchdog *cluster.Watchdog
+	wasLeader := false
 	for {
 		select {
 		case isElected := <-electedCh:
 			if isElected {
 				log.Info("Leader Election: Cluster leadership acquired")
+				wasLeader = true
 				watchdog = cluster.NewWatchdog(cl)
 				server.SetHandler(primary)
 			} else {
 				log.Info("Leader Election: Cluster leadership lost")
 				cl.UnregisterEventHandler(watchdog)
 				server.SetHandler(replica)
+				if wasLeader {
```

Comment:
Definitely. Still, I really want to find a good way to avoid these 20 second long delays when we lose connection to the KV store in a single-node environment.